htdocs/core/login/functions_openid_connect.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/* Copyright (C) 2022 Jeritiana Ravelojaona <jeritiana.rav@smartone.ai>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/**
+ *      \file       htdocs/core/login/functions_openid_connect.php
+ *      \ingroup    core
+ *      \brief      OpenID Connect: Authorization Code flow authentication
+ */
+
+include_once DOL_DOCUMENT_ROOT.'/core/lib/geturl.lib.php';
+
+/**
+ * Check validity of user/password/entity
+ * If test is ko, reason must be filled into $_SESSION["dol_loginmesg"]
+ *
+ * @param	string	$usertotest		Login
+ * @param	string	$passwordtotest	Password
+ * @param   int		$entitytotest   Number of instance (always 1 if module multicompany not enabled)
+ * @return	string					Login if OK, '' if KO
+ */
+function check_user_password_openid_connect($usertotest, $passwordtotest, $entitytotest)
+{
+	global $conf;
+
+	dol_syslog("functions_openid_connect::check_user_password_openid_connect");
+
+	$login = '';
+
+	// Step 1 is done by user: request an authorization code
+
+	if (GETPOSTISSET('username')) {
+		// OIDC does not require credentials here: pass on to next auth handler
+		$_SESSION["dol_loginmesg"] = "Not an OpenID Connect flow";
+		dol_syslog("functions_openid_connect::check_user_password_openid_connect not an OIDC flow");
+	} elseif (GETPOSTISSET['code']) {
+		$auth_code = GETPOST('code', 'aZ09');
+		dol_syslog("functions_openid_connect::check_user_password_openid_connect code=".$auth_code);
+
+		// Step 2: turn the authorization code into an access token, using client_secret
+		$auth_param = [
+			'grant_type'    => 'authorization_code',
+			'client_id'     => $conf->global->MAIN_AUTHENTICATION_OIDC_CLIENT_ID,
+			'client_secret' => $conf->global->MAIN_AUTHENTICATION_OIDC_CLIENT_SECRET,
+			'code'          => $auth_code,
+						'redirect_uri'  => $conf->global->MAIN_AUTHENTICATION_OIDC_REDIRECT_URL
+		];
+
+		$token_response = getURLContent($conf->global->MAIN_AUTHENTICATION_OIDC_TOKEN_URL, 'POST', http_build_query($auth_param));
+		$token_content = json_decode($token_response['content']);
+		dol_syslog("functions_openid_connect::check_user_password_openid_connect /token=".print_r($token_response, true), LOG_DEBUG);
```

Comment:
$token_response = getURLContent($conf->global->MAIN_AUTHENTICATION_OIDC_TOKEN_URL, 'POST', http_build_query($auth_param),1, array(),array('http', 'https'),2);
=> allow local identity provider (in my case, I use virtual machines on the same network so 192.168.xx.xx and it's blocked)